Remove constexpr from ringOrder

davidlt opened the pull request:
ICC segfaults due to to `constexpr` on `ringOrder`. This is reported and
due to be fixed in forthcoming ICC releases (2016).

Removal of `constexpr` did not change assembly in any significant,
actually compiler generated 5 instructions less for this function.

Signed-off-by: David Abdurachmanov David.Abdurachmanov@cern.ch

cmsbuild commented:
A new Pull Request was created by @davidlt  for CMSSW_7_6_X.

Remove constexpr from ringOrder

It involves the following packages:

RecoTracker/TkDetLayers


@cmsbuild, @cvuosalo, @slava77 can you please review it and eventually sign? Thanks.
@ghellwig, @makortel, @GiacomoSguazzoni, @rovere, @VinInn, @mschrode, @istaslis, @gpetruc, @cerati, @dgulhan this is something you requested to watch as well.
You can sign-off by replying to this message having '+1' in the first line of your reply.
You can reject by replying  to this message having '-1' in the first line of your reply.
If you are a L2 or a release manager you can ask for tests by saying 'please test' in the first line of a comment.
@degano you are the release manager for this.
You can merge this pull request by typing 'merge' in the first line of your comment.

slava77 commented on RecoTracker/TkDetLayers/src/Phase2OTECRingedLayer.cc:
```diff
@@ -128,7 +128,7 @@ Phase2OTECRingedLayer::groupedCompatibleDetsV( const TrajectoryStateOnSurface& s
   }
 
   //order is odd rings in front of even rings
-  constexpr int ringOrder[NOTECRINGS]{0,1,0,1,0,1,0,1,0,1,0,1,0,1,0};
```
can we keep "const" here?
This is logically what it was before the change proposed here.

davidlt replied:
Yes, I don't know how that got away from me.

VinInn commented:
Can we stop modifying production code to support development platform used only for R&D?
We need a policy here. I strongly oppose to change any code with the only purpose to support R&D on HW/Software that has close to negligible chance to become a production platform in the medium future.

davidlt commented:
@VinInn I am happy to add ifdef here if required. It's L2 decision. But technically there seems to be no gain in `constexpr` here.

cmsbuild commented:
Pull request #11848 was updated. @cmsbuild, @cvuosalo, @slava77 can you please check and sign again.

davidlange6 commented:
if I understand correctly that we are patching around an ICC bug (which indeed is not likely a production cms platform..) - an ifdef sounds like the best approach.

Commit:
Remove constexpr from ringOrder

ICC segfaults due to to `constexpr` on `ringOrder`. This is reported and
due to be fixed in forthcoming ICC releases (2016).

Removal of `constexpr` did not change assembly in any significant,
actually compiler generated 5 instructions less for this function.

Signed-off-by: David Abdurachmanov <David.Abdurachmanov@cern.ch>

cmsbuild commented:
Pull request #11848 was updated. @cmsbuild, @cvuosalo, @slava77 can you please check and sign again.

davidlt commented:
I added ifdefs, but again `constexpr` aren't giving any benefits here.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.
JENKINS_TEST_URL

cmsbuild commented:
+1
Tested at: 9d7617071e70e8f07dd60e6dfe4c326ef3ca5df9
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11848/8958/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11848/8958/summary.html

cvuosalo commented:
+1

For #11848 9d7617071e70e8f07dd60e6dfe4c326ef3ca5df9

Adding #ifdef for a special R&D compiler that has a bug with constexpr to allow it to compile CMSSW. This compiler bug might get fixed in 2016.

There should be no change in run-time behavior.

The code changes are satisfactory, and standard compilation is successful. Jenkins [tests](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11848/8958/summary.html) against baseline CMSSW_7_6_X_2015-10-16-1100 show no significant differences, as expected.

cvuosalo commented:
@davidlt:
It would be nice if you could remove these #ifdefs next year when you get a new version of ICC without the bug.

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_7_6_X IBs (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @degano, @smuzaffar

davidlange6 commented:
+1
